tree/ntuple/v7/src/RPageStorage.cxx

```diff
@@ -240,52 +240,57 @@ void ROOT::Experimental::Internal::RPageSource::UnzipClusterImpl(RCluster *clust
-      const auto &columnDesc = descriptorGuard->GetColumnDescriptor(columnId);
-
-      allElements.emplace_back(RColumnElementBase::Generate(columnDesc.GetType()));
-
-      const auto &pageRange = clusterDescriptor.GetPageRange(columnId);
-      std::uint64_t pageNo = 0;
-      std::uint64_t firstInPage = 0;
-      for (const auto &pi : pageRange.fPageInfos) {
-         ROnDiskPage::Key key(columnId, pageNo);
-         auto onDiskPage = cluster->GetOnDiskPage(key);
-         RSealedPage sealedPage;
-         sealedPage.SetNElements(pi.fNElements);
-         sealedPage.SetHasChecksum(pi.fHasChecksum);
-         sealedPage.SetBufferSize(pi.fLocator.fBytesOnStorage + pi.fHasChecksum * kNBytesPageChecksum);
-         sealedPage.SetBuffer(onDiskPage->GetAddress());
-         R__ASSERT(onDiskPage && (onDiskPage->GetSize() == sealedPage.GetBufferSize()));
-
-         auto taskFunc = [this, columnId, clusterId, firstInPage, sealedPage, element = allElements.back().get(),
-                          &foundChecksumFailure,
-                          indexOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex]() {
-            auto rv = UnsealPage(sealedPage, *element, columnId);
-            if (!rv) {
-               foundChecksumFailure = true;
-               return;
-            }
-            auto newPage = rv.Unwrap();
-            fCounters->fSzUnzip.Add(element->GetSize() * sealedPage.GetNElements());
+      // By the time we unzip a cluster, the set of active columns may have already changed wrt. to the moment when
+      // we requested reading the cluster. That doesn't matter much, we simply decompress what is now in the list
+      // of active columns.
+      if (!fActivePhysicalColumns.HasColumnInfos(columnId))
+         continue;
+      const auto &columnInfos = fActivePhysicalColumns.GetColumnInfos(columnId);
+
+      for (const auto &info : columnInfos) {
+         const auto elementId = info.fElementId;
+
+         const auto &pageRange = clusterDescriptor.GetPageRange(columnId);
+         std::uint64_t pageNo = 0;
+         std::uint64_t firstInPage = 0;
+         for (const auto &pi : pageRange.fPageInfos) {
+            ROnDiskPage::Key key(columnId, pageNo);
+            auto onDiskPage = cluster->GetOnDiskPage(key);
+            RSealedPage sealedPage;
+            sealedPage.SetNElements(pi.fNElements);
+            sealedPage.SetHasChecksum(pi.fHasChecksum);
+            sealedPage.SetBufferSize(pi.fLocator.fBytesOnStorage + pi.fHasChecksum * kNBytesPageChecksum);
+            sealedPage.SetBuffer(onDiskPage->GetAddress());
+            R__ASSERT(onDiskPage && (onDiskPage->GetSize() == sealedPage.GetBufferSize()));
+
+            auto taskFunc = [this, columnId, clusterId, firstInPage, sealedPage, elementId, &foundChecksumFailure,
+                             indexOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex]() {
+               auto element = GenerateColumnElement(elementId);
```

Comment:
This generates a column element per page, while we can share it across a column info. I think this (relatively simple) optimization should be retained from the current implementation (cf `allElements`)